nsv13/code/modules/research/astrometrics.dm

```diff
@@ -0,0 +1,129 @@
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		var/datum/asset/assets = get_asset_datum(/datum/asset/simple/starmap)
+		assets.send(user)
+		ui = new(user, src, ui_key, "astrometrics", name, 800, 660, master_ui, state)
+		ui.open()
+
+/**
+Clean override of the navigation computer to provide scan functionality.
+*/
+
+/obj/machinery/computer/ship/navigation/astrometrics/ui_data(mob/user)
+	var/list/data = ..()
+	var/list/info = SSstar_system.ships[linked]
+	var/datum/star_system/current_system = info["current_system"]
+	if(scan_target)
+		data["scan_target"] = scan_target.name
+	if(screen == 2) // Here's where the magic happens.
+		data["star_id"] = "\ref[selected_system]"
+		data["star_name"] = selected_system.name
+		data["alignment"] = capitalize(selected_system.alignment)
+		data["scanned"] = FALSE
+		if(info["current_system"])
+			data["star_dist"] = info["current_system"].dist(selected_system)
+			data["can_scan"] = is_in_range(current_system, selected_system)
+			data["can_cancel"] = (scan_target) ? TRUE : FALSE
+		if(LAZYFIND(scanned, selected_system.name)) //If we've scanned this one before, get me the list of its anomalies.
+			data["anomalies"] = selected_system.get_info()
+			data["scanned"] = TRUE
+	data["scan_progress"] = scan_progress
+	data["scan_goal"] = scan_goal
+	return data
+
+/obj/machinery/computer/ship/navigation/astrometrics/is_in_range(datum/star_system/current_system, datum/star_system/system)
+	return current_system && current_system.dist(system) <= max_range
+
+/obj/machinery/computer/ship/navigation/astrometrics/is_visited(datum/star_system/system)
+	return LAZYFIND(scanned, system.name)
+
+/obj/machinery/computer/ship/navigation/astrometrics/ui_act(action, params, datum/tgui/ui)
+	if(..())
+		return
+	if(!has_overmap())
+		return
+	switch(action)
+		if("scan")
+			scan_goal = initial(scan_goal)
+			scan_target = selected_system
+			say("Initiating scan of: [scan_target]")
+			playsound(src, 'nsv13/sound/voice/scan_start.wav', 100, FALSE)
+			radio.talk_into(src, "Initiating scan of: [scan_target]", channel)
+		if("scan_anomaly")
+			var/obj/effect/overmap_anomaly/target = locate(params["anomaly_id"])
+			if(!istype(target))
+				return
+			scan_goal = initial(scan_goal) / 2
+			scan_target = target
+			say("Initiating scan of: [scan_target]")
+			radio.talk_into(src, "Initiating scan of: [scan_target]", channel)
+		if("cancel_scan")
+			say("Scan of [scan_target] cancelled!")
+			playsound(src, 'nsv13/sound/voice/scanning_cancelled.wav', 100, FALSE)
+			radio.talk_into(src, "Scan of [scan_target] cancelled!", channel)
+			scan_progress = 0
+			scan_target = null
+
+/obj/machinery/computer/ship/navigation/astrometrics/process()
+	if(scan_target)
+		scan_progress += 1 SECONDS
```

Comment:
Why do this via a process instead of a timer? It seems a bit wasteful to have it doing it constantly than just doing one at the start then another at the end of the time.

Reply:
So you can see the progress of the scan.

Reply:
I believe there is a way that was discussed on TG of knowing how long until its finished while its a timer.